crates/router/src/db/mandate.rs

```diff
@@ -108,46 +111,149 @@ impl MandateInterface for Store {
+                    storage::MandateUpdate::StatusUpdate { mandate_status } => {
+                        mandate.mandate_status = mandate_status;
+                    }
+                    storage::MandateUpdate::CaptureAmountUpdate { amount_captured } => {
+                        mandate.amount_captured = amount_captured;
+                    }
+                    storage::MandateUpdate::ConnectorReferenceUpdate {
+                        connector_mandate_ids,
+                    } => {
+                        mandate.connector_mandate_ids = connector_mandate_ids;
+                    }
+                }
+                Ok(mandate.clone())
+            }
+            None => {
+                Err(errors::StorageError::ValueNotFound("mandate not found".to_string()).into())
+            }
+        }
     }
 
     async fn find_mandates_by_merchant_id(
         &self,
-        _merchant_id: &str,
-        _mandate_constraints: api_models::mandates::MandateListConstraints,
+        merchant_id: &str,
+        mandate_constraints: api_models::mandates::MandateListConstraints,
     ) -> CustomResult<Vec<storage::Mandate>, errors::StorageError> {
-        // [#172]: Implement function for `MockDb`
-        Err(errors::StorageError::MockDbError)?
+        let mandates = self.mandates.lock().await;
+        let mandates_iter = mandates.iter().filter(|mandate| {
+            let mut checker = mandate.merchant_id == merchant_id;
+            if let Some(created_time) = mandate_constraints.created_time {
+                checker &= mandate.created_at == created_time;
+            }
+            if let Some(created_time_lt) = mandate_constraints.created_time_lt {
+                checker &= mandate.created_at < created_time_lt;
+            }
+            if let Some(created_time_gt) = mandate_constraints.created_time_gt {
+                checker &= mandate.created_at > created_time_gt;
+            }
+            if let Some(created_time_lte) = mandate_constraints.created_time_lte {
+                checker &= mandate.created_at <= created_time_lte;
+            }
+            if let Some(created_time_gte) = mandate_constraints.created_time_gte {
+                checker &= mandate.created_at >= created_time_gte;
+            }
+            if let Some(connector) = &mandate_constraints.connector {
+                checker &= mandate.connector == *connector;
+            }
+            if let Some(mandate_status) = mandate_constraints.mandate_status {
+                let storage_mandate_status: storage_models::enums::MandateStatus =
+                    mandate_status.foreign_into();
+                checker &= mandate.mandate_status == storage_mandate_status;
+            }
+            checker
+        });
+
+        let mandates: Vec<storage::Mandate> = if let Some(limit) = mandate_constraints.limit {
+            #[allow(clippy::as_conversions)]
+            mandates_iter.take(limit as usize).cloned().collect()
```

Comment:
can we have a bit of runtime safety here, consider this
limit is 
`Option<i64>` -> `i64`
to convert `i64` to `usize` we need to handle negative numbers
```
x if x < 0 => 0,
x => x as usize,
```

Reply:
Should we consider using `usize::MAX` if `limit` is negative?

Reply:
if the limit is not passed, then we are displaying the same behaviour as `usize::MAX`, so I think in case of 0 we should return no entires